File: android/joinery-android-dnsfilter/src/main/java/com/getjoinery/dnsfilter/DnsFilterApi.kt
package com.getjoinery.dnsfilter

import com.getjoinery.android.ApiClient
import com.getjoinery.android.JoineryApiError
import com.getjoinery.android.JsonValue

/**
 * Thin typed face over the `dns_filtering/` action namespace
 * (`POST /api/v1/action/dns_filtering/{action}`). Every call rides the app's
 * session key through ApiClient; tier gating, ownership, and save semantics are
 * entirely server-side — the same logic functions the web editor calls — so
 * this layer holds no policy, only shapes.
 */
// `open` so a store test can subclass and control timing; production always
// uses the concrete methods below.
open class DnsFilterApi(val client: ApiClient) {

    private fun ns(name: String): String = "dns_filtering/$name"

    private fun emptyBody(): JsonValue = JsonValue.Obj(emptyList())

    // MARK: Reads

    /** List the user's devices with DoH/DoT endpoints, per-block summaries, and
     *  the merged hard-block hostname list. */
    open suspend fun devices(): List<DnsDevice> {
        val envelope = client.submitAction(ns("devices"), emptyBody())
        return (envelope["data"]?.get("devices")?.arrayValue ?: emptyList())
            .mapNotNull { DnsDevice.from(it) }
    }

    /** Tier name, the five feature flags, and device count vs. limit. */
    suspend fun accountSummary(): DnsAccountSummary {
        val envelope = client.submitAction(ns("account_summary"), emptyBody())
        return DnsAccountSummary.from(envelope["data"]) ?: throw JoineryApiError.Malformed
    }

    /** The filter/service catalog — static per deployment, cache client-side. */
    suspend fun catalog(): DnsCatalog {
        val envelope = client.submitAction(ns("catalog"), emptyBody())
        return DnsCatalog.from(envelope["data"]) ?: throw JoineryApiError.Malformed
    }

    /** Read one block's full contents (`scheduled_block_edit` with no `action`
     *  key). [deviceId] is required; [blockId] selects the always-on or a
     *  specific scheduled block. */
    suspend fun blockContents(deviceId: Int, blockId: Int): DnsBlockContents {
        val envelope = client.submitAction(
            ns("scheduled_block_edit"),
            JsonValue.obj(
                "device_id" to JsonValue.Num(deviceId.toDouble()),
                "block_id" to JsonValue.Num(blockId.toDouble()),
            ),
        )
        return DnsBlockContents.from(envelope["data"]) ?: throw JoineryApiError.Malformed
    }

    // MARK: Category / service toggles (save-on-change)

    /** Set or clear one category filter on a block. `action` null == Allow,
     *  which the server writes as "no row" (the resolver-merge invariant —
     *  "Allow means no row"). Blocking writes an `action=0` row. */
    suspend fun setFilter(blockId: Int, key: String, action: Int?) =
        setRule(blockId, "filter", key, action)

    /** Set or clear one service toggle on a block. */
    suspend fun setService(blockId: Int, key: String, action: Int?) =
        setRule(blockId, "service", key, action)

    private suspend fun setRule(blockId: Int, type: String, key: String, action: Int?) {
        client.submitAction(
            ns("block_filter_set"),
            JsonValue.obj(
                "block_id" to JsonValue.Num(blockId.toDouble()),
                "type" to JsonValue.Str(type),
                "key" to JsonValue.Str(key),
                // Empty string removes the row; "0" blocks, "1" allows.
                "action" to JsonValue.Str(action?.toString() ?: ""),
            ),
        )
    }

    // MARK: Custom domain rules

    /** Add a custom domain rule. [hardBlock] (block-action + always-on block
     *  only, server-enforced) marks it for the strict-mode tunnel. Rides the
     *  `scrolldaddy_custom_rules` gate. */
    suspend fun addDomainRule(blockId: Int, hostname: String, action: Int, hardBlock: Boolean = false): DnsDomainRule {
        val envelope = client.submitAction(
            ns("block_rule_add"),
            JsonValue.obj(
                "block_id" to JsonValue.Num(blockId.toDouble()),
                "hostname" to JsonValue.Str(hostname),
                "action" to JsonValue.Num(action.toDouble()),
                "hard_block" to JsonValue.Bool(hardBlock),
            ),
        )
        return DnsDomainRule.from(envelope["data"]) ?: throw JoineryApiError.Malformed
    }

    suspend fun deleteDomainRule(ruleId: Int) {
        client.submitAction(
            ns("block_rule_delete"),
            JsonValue.obj("rule_id" to JsonValue.Num(ruleId.toDouble())),
        )
    }

    // MARK: Devices

    /**
     * Register a device (omit `device_id`). The create path is keyed on
     * `device_name` and reads `device_type` / `sdd_timezone` /
     * `sdd_allow_device_edits` (device_edit_logic + SdDevice::createDevice).
     * In API context the server returns the new device directly (device_edit
     * server contract), so this parses and returns it; a null return means the
     * response wasn't a device (e.g. a server without the API-create contract),
     * and the caller falls back to a before/after device-list diff.
     */
    open suspend fun createDevice(name: String, deviceType: String, timezone: String): DnsDevice? {
        val envelope = client.submitAction(
            ns("device_edit"),
            JsonValue.obj(
                "device_name" to JsonValue.Str(name),
                "device_type" to JsonValue.Str(deviceType),
                "sdd_timezone" to JsonValue.Str(timezone),
                // The create path reads this unguarded; the app always allows edits.
                "sdd_allow_device_edits" to JsonValue.Str("1"),
            ),
        )
        return DnsDevice.from(envelope["data"])
    }
}

File: ios/joinery-kit/Sources/JoineryDNSFilterKit/DNSFilterAPI.swift
import Foundation
import JoineryKit

/// Thin typed face over the `dns_filtering/` action namespace
/// (`POST /api/v1/action/dns_filtering/{action}`). Every call rides the app's
/// session key through APIClient; tier gating, ownership, and save semantics
/// are entirely server-side — the same logic functions the web editor calls —
/// so this layer holds no policy, only shapes.
public struct DNSFilterAPI: Sendable {
    let client: APIClient

    public init(client: APIClient) {
        self.client = client
    }

    private func ns(_ name: String) -> String { "dns_filtering/\(name)" }

    // MARK: Reads

    /// List the user's devices with DoH/DoT endpoints, per-block summaries,
    /// and the merged hard-block hostname list.
    public func devices() async throws -> [DNSDevice] {
        let envelope = try await client.submitAction(ns("devices"), body: .object([]))
        return (envelope["data"]?["devices"]?.arrayValue ?? []).compactMap(DNSDevice.init(json:))
    }

    /// Tier name, the five feature flags, and device count vs. limit.
    public func accountSummary() async throws -> DNSAccountSummary {
        let envelope = try await client.submitAction(ns("account_summary"), body: .object([]))
        guard let summary = DNSAccountSummary(data: envelope["data"]) else {
            throw JoineryAPIError.malformedResponse
        }
        return summary
    }

    /// The filter/service catalog — static per deployment, cache client-side.
    public func catalog() async throws -> DNSCatalog {
        let envelope = try await client.submitAction(ns("catalog"), body: .object([]))
        guard let catalog = DNSCatalog(data: envelope["data"]) else {
            throw JoineryAPIError.malformedResponse
        }
        return catalog
    }

    /// Read one block's full contents (`scheduled_block_edit` with no `action`
    /// key). `device_id` is required; `block_id` selects the always-on or a
    /// specific scheduled block.
    public func blockContents(deviceID: Int, blockID: Int) async throws -> DNSBlockContents {
        let envelope = try await client.submitAction(ns("scheduled_block_edit"), body: .object([
            (key: "device_id", value: .number(Double(deviceID))),
            (key: "block_id", value: .number(Double(blockID))),
        ]))
        guard let contents = DNSBlockContents(data: envelope["data"]) else {
            throw JoineryAPIError.malformedResponse
        }
        return contents
    }

    // MARK: Category / service toggles (save-on-change)

    /// Set or clear one category filter on a block. `action` nil == Allow,
    /// which the server writes as "no row" (the resolver-merge invariant —
    /// "Allow means no row"). Blocking writes an `action=0` row.
    public func setFilter(blockID: Int, key: String, action: Int?) async throws {
        try await setRule(blockID: blockID, type: "filter", key: key, action: action)
    }

    /// Set or clear one service toggle on a block.
    public func setService(blockID: Int, key: String, action: Int?) async throws {
        try await setRule(blockID: blockID, type: "service", key: key, action: action)
    }

    private func setRule(blockID: Int, type: String, key: String, action: Int?) async throws {
        _ = try await client.submitAction(ns("block_filter_set"), body: .object([
            (key: "block_id", value: .number(Double(blockID))),
            (key: "type", value: .string(type)),
            (key: "key", value: .string(key)),
            // Empty string removes the row; "0" blocks, "1" allows.
            (key: "action", value: .string(action.map { String($0) } ?? "")),
        ]))
    }

    // MARK: Custom domain rules

    /// Add a custom domain rule. `hardBlock` (block-action + always-on block
    /// only, server-enforced) marks it for the strict-mode tunnel. Rides the
    /// `scrolldaddy_custom_rules` gate.
    @discardableResult
    public func addDomainRule(blockID: Int, hostname: String, action: Int, hardBlock: Bool = false) async throws -> DNSDomainRule {
        let envelope = try await client.submitAction(ns("block_rule_add"), body: .object([
            (key: "block_id", value: .number(Double(blockID))),
            (key: "hostname", value: .string(hostname)),
            (key: "action", value: .number(Double(action))),
            (key: "hard_block", value: .bool(hardBlock)),
        ]))
        guard let rule = DNSDomainRule(json: envelope["data"]) else {
            throw JoineryAPIError.malformedResponse
        }
        return rule
    }

    public func deleteDomainRule(ruleID: Int) async throws {
        _ = try await client.submitAction(ns("block_rule_delete"), body: .object([
            (key: "rule_id", value: .number(Double(ruleID))),
        ]))
    }

    // MARK: Devices

    /// Register a device (omit `device_id`). The create path is keyed on
    /// `device_name` and reads `device_type` / `sdd_timezone` /
    /// `sdd_allow_device_edits` (device_edit_logic + SdDevice::createDevice),
    /// then redirects — it returns no id, so the caller diffs the device list
    /// to find the new row. Rides the `scrolldaddy_max_devices` gate.
    public func createDevice(name: String, deviceType: String, timezone: String) async throws {
        _ = try await client.submitAction(ns("device_edit"), body: .object([
            (key: "device_name", value: .string(name)),
            (key: "device_type", value: .string(deviceType)),
            (key: "sdd_timezone", value: .string(timezone)),
            // The create path reads this unguarded; the app always allows edits.
            (key: "sdd_allow_device_edits", value: .string("1")),
        ]))
    }
}
